pkg/storage/cache/object.go

```diff
@@ -0,0 +1,250 @@
+package cache
+
+import (
+	"context"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/jmoiron/sqlx"
+	"github.com/juju/loggo"
+	v1 "github.com/zhihu/cmdb/pkg/api/v1"
+	"github.com/zhihu/cmdb/pkg/model"
+	"github.com/zhihu/cmdb/pkg/model/objects"
+	"github.com/zhihu/cmdb/pkg/model/typetables"
+	"github.com/zhihu/cmdb/pkg/storage"
+	"github.com/zhihu/cmdb/pkg/storage/cdc"
+)
+
+var log = loggo.GetLogger("cache")
+
+type Objects struct {
+	name           string
+	id             int
+	loaded         bool
+	memory         objects.Database
+	handlers       []storage.FilterWatcher
+	typ            *Types
+	objects        map[int]*v1.Object
+	mutex          sync.RWMutex
+	bufferedEvents [][]cdc.Event
+}
+
+func NewObjects(typ *Types, id int, name string) *Objects {
+	return &Objects{
+		name:    name,
+		id:      id,
+		typ:     typ,
+		objects: map[int]*v1.Object{},
+		mutex:   sync.RWMutex{},
+	}
+}
+
+func (t *Objects) ResetBuffer() {
+	t.bufferedEvents = nil
+}
+
+func (t *Objects) LoadData(ctx context.Context, db *sqlx.DB) error {
+	tx, err := db.Beginx()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+	var list []*model.Object
+	err = tx.SelectContext(ctx, &list, `select * from object where type_id = ? and delete_time is null`, t.id)
+	if err != nil {
+		return err
+	}
+	var metas []*model.ObjectMetaValue
+	err = tx.SelectContext(ctx, &metas, `select m.* from object_meta_value m left join object o on m.object_id = o.id where o.type_id = ? and o.delete_time is null and m.delete_time is null `, t.id)
+	if err != nil {
+		return err
+	}
+	_ = tx.Commit()
+	t.mutex.Lock()
+	t.memory.Init()
+	for _, object := range list {
+		t.memory.InsertObject(object)
+	}
+	for _, meta := range metas {
+		t.memory.InsertObjectMetaValue(meta)
+	}
+	t.loaded = true
+	t.objects = map[int]*v1.Object{}
+	for id, object := range t.memory.ObjectTable.ID {
+		t.objects[id.ID] = t.convert(object)
+	}
+	t.mutex.Unlock()
+
+	return nil
+}
+
+func (t *Objects) RemoveFilterWatcher(f storage.FilterWatcher) {
+	t.mutex.Lock()
+	var nHandlers []storage.FilterWatcher
```

Comment:
pre allocate a large enough buffer or simply update array in place can save unnecessary allocations.